transmart-server/grails-app/services/org/transmart/server/subscription/QuerySetSubscriptionMailService.groovy

```diff
@@ -0,0 +1,139 @@
+import org.transmartproject.core.users.UsersResource
+import org.transmartproject.core.userquery.SetType
+
+/**
+ * Generates and sends a daily or weekly subscription email for each user
+ * that subscribed for a query data result updates.
+ *
+ * The DSL for sending emails is provided by the Grails mail plugin.
+ * See Grails Mail Plugin documentation for more information:
+ * http://gpc.github.io/grails-mail/
+ *
+ */
+@Slf4j
+class QuerySetSubscriptionMailService {
+
+    def grailsApplication
+
+    MailService mailService
+
+    @Autowired
+    UsersResource usersResource
+
+    @Autowired
+    UserQuerySetResource userQueryDiffResource
+
+    private final static String NEW_LINE = "\n"
+
+    /**
+     * Creates and sends a daily or weekly email for each user having an email specified.
+     *
+     * The email is sent only when there are changes for one or more of the results of queries the user subscribed for.
+     * A query result is changed when there is an object ID in the result that wasn't there the last time it was run,
+     * or when an object ID isn't there which was there the last time it was run.
+     *
+     */
+    def run(SubscriptionFrequency frequency) {
+        List<User> users = getUsers()
+        for (user in users) {
+            String title = "Test email for query subscription"
+            String emailBody = generateEmail(user.username, frequency)
+            if(emailBody) {
+                sendEmail(user.email, title, emailBody)
+            }
+        }
+    }
+
+    /**
+     * Generates an email for specific user with data updates for a query the user is subscribed for.
+     *
+     * The email contains a list of each query with changed results with:
+     * - a name of the query,
+     * - list of added and removed ids of objects that the query relates to
+     * - over what period the change was
+     *
+     * @param username
+     * @param freq
+     * @return The body of the email
+     *
+     */
+    private String generateEmail(String username, SubscriptionFrequency frequency) {
+
+        Integer maxNumberOfSets = grailsApplication.config.org.transmart.server.subscription.maxNumberOfSets
+        def currentDate = new Date()
+
+        List<UserQuerySetChangesRepresentation> querySetChanges = userQueryDiffResource
+                .getQueryChangeHistoryByUsernameAndFrequency(frequency, username, maxNumberOfSets)
+
+        if (querySetChanges.size() > 0) {
+
+            StringBuilder textStringBuilder = new StringBuilder()
+            textStringBuilder.append("Generated as per day ${currentDate.format("d.' of 'MMMM Y h:mm aa z")}")
+            textStringBuilder.append(NEW_LINE)
+            textStringBuilder.append("List of updated query results:")
+            textStringBuilder.append(NEW_LINE)
+
+            def patientSets = querySetChanges.findAll { it.setType == SetType.PATIENT }?.sort { it.queryId }
+            if (patientSets.size() == 0) {
+                return null
```

Comment:
Try don't return `null`. In this particular example, there is no other type then patient set.